File: flowstudio-frontend/src/services/websocket.ts
/**
 * FlowStudio WebSocket Service
 * Handles real-time communication for execution updates
 */

export interface ExecutionUpdate {
  type: 'execution_update';
  execution_id: string;
  update_type: 'status' | 'progress' | 'component' | 'log' | 'error' | 'result';
  data: any;
  timestamp: string;
}

export interface WebSocketMessage {
  type: string;
  [key: string]: any;
}

export type ExecutionUpdateHandler = (update: ExecutionUpdate) => void;
export type ConnectionHandler = (connected: boolean) => void;

class WebSocketService {
  private ws: WebSocket | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;
  private executionHandlers: Map<string, Set<ExecutionUpdateHandler>> = new Map();
  private connectionHandlers: Set<ConnectionHandler> = new Set();
  private isConnecting = false;

  /**
   * Connect to WebSocket server
   */
  connect(token: string): Promise<void> {
    if (this.ws?.readyState === WebSocket.OPEN || this.isConnecting) {
      return Promise.resolve();
    }

    this.isConnecting = true;

    return new Promise((resolve, reject) => {
      try {
        const wsUrl = `ws://localhost:8003/api/fs/ws?token=${encodeURIComponent(token)}`;
        this.ws = new WebSocket(wsUrl);

        this.ws.onopen = () => {
          console.log('WebSocket connected');
          this.isConnecting = false;
          this.reconnectAttempts = 0;
          this.notifyConnectionHandlers(true);
          resolve();
        };

        this.ws.onmessage = (event) => {
          try {
            const message: WebSocketMessage = JSON.parse(event.data);
            this.handleMessage(message);
          } catch (error) {
            console.error('Failed to parse WebSocket message:', error);
          }
        };

        this.ws.onclose = (event) => {
          console.log('WebSocket disconnected:', event.code, event.reason);
          this.isConnecting = false;
          this.notifyConnectionHandlers(false);
          
          // Attempt to reconnect if not a manual disconnect
          if (event.code !== 1000 && this.reconnectAttempts < this.maxReconnectAttempts) {
            this.scheduleReconnect(token);
          }
        };

        this.ws.onerror = (error) => {
          console.error('WebSocket error:', error);
          this.isConnecting = false;
          reject(new Error('WebSocket connection failed'));
        };

      } catch (error) {
        this.isConnecting = false;
        reject(error);
      }
    });
  }

  /**
   * Disconnect from WebSocket server
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close(1000, 'Manual disconnect');
      this.ws = null;
    }
    this.executionHandlers.clear();
    this.connectionHandlers.clear();
  }

  /**
   * Subscribe to execution updates
   */
  subscribeToExecution(executionId: string, handler: ExecutionUpdateHandler): () => void {
    if (!this.executionHandlers.has(executionId)) {
      this.executionHandlers.set(executionId, new Set());
      
      // Send subscription message to server
      this.sendMessage({
        type: 'subscribe_execution',
        execution_id: executionId
      });
    }

    this.executionHandlers.get(executionId)!.add(handler);

    // Return unsubscribe function
    return () => {
      const handlers = this.executionHandlers.get(executionId);
      if (handlers) {
        handlers.delete(handler);
        if (handlers.size === 0) {
          this.executionHandlers.delete(executionId);
          
          // Send unsubscription message to server
          this.sendMessage({
            type: 'unsubscribe_execution',
            execution_id: executionId
          });
        }
      }
    };
  }

  /**
   * Subscribe to connection status changes
   */
  onConnectionChange(handler: ConnectionHandler): () => void {
    this.connectionHandlers.add(handler);
    
    // Return unsubscribe function
    return () => {
      this.connectionHandlers.delete(handler);
    };
  }

  /**
   * Get current connection status
   */
  isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }

  /**
   * Send a message to the server
   */
  private sendMessage(message: any): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(message));
    } else {
      console.warn('WebSocket not connected, cannot send message:', message);
    }
  }

  /**
   * Handle incoming WebSocket messages
   */
  private handleMessage(message: WebSocketMessage): void {
    switch (message.type) {
      case 'execution_update':
        this.handleExecutionUpdate(message as ExecutionUpdate);
        break;
        
      case 'connection_established':
        console.log('WebSocket connection established');
        break;
        
      case 'subscription_confirmed':
        console.log('Subscription confirmed for execution:', message.execution_id);
        break;
        
      case 'unsubscription_confirmed':
        console.log('Unsubscription confirmed for execution:', message.execution_id);
        break;
        
      case 'error':
        console.error('WebSocket error message:', message.message);
        break;
        
      default:
        console.log('Unknown WebSocket message type:', message.type);
    }
  }

  /**
   * Handle execution update messages
   */
  private handleExecutionUpdate(update: ExecutionUpdate): void {
    const handlers = this.executionHandlers.get(update.execution_id);
    if (handlers) {
      handlers.forEach(handler => {
        try {
          handler(update);
        } catch (error) {
          console.error('Error in execution update handler:', error);
        }
      });
    }
  }

  /**
   * Notify connection handlers of status change
   */
  private notifyConnectionHandlers(connected: boolean): void {
    this.connectionHandlers.forEach(handler => {
      try {
        handler(connected);
      } catch (error) {
        console.error('Error in connection handler:', error);
      }
    });
  }

  /**
   * Schedule a reconnection attempt
   */
  private scheduleReconnect(token: string): void {
    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);
    
    console.log(`Scheduling WebSocket reconnect attempt ${this.reconnectAttempts} in ${delay}ms`);
    
    setTimeout(() => {
      if (this.reconnectAttempts <= this.maxReconnectAttempts) {
        this.connect(token).catch(error => {
          console.error('WebSocket reconnect failed:', error);
        });
      }
    }, delay);
  }

  /**
   * Send ping message for heartbeat
   */
  ping(): void {
    this.sendMessage({
      type: 'ping',
      timestamp: new Date().toISOString()
    });
  }
}

// Export singleton instance
export const webSocketService = new WebSocketService();

// Auto-connect when authentication token is available
export const connectWebSocket = async (): Promise<void> => {
  const token = localStorage.getItem('access_token');
  if (token && !webSocketService.isConnected()) {
    try {
      await webSocketService.connect(token);
    } catch (error) {
      console.error('Failed to connect WebSocket:', error);
    }
  }
};

export default webSocketService;

File: flowstudio-backend/app/api/websocket.py
"""
FlowStudio WebSocket API
Handles real-time communication for execution updates
"""
import json
import logging
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Depends, Query
from typing import Optional

from ..services.websocket_manager import connection_manager
from ..api.deps_external_auth import verify_token_with_auth_server

logger = logging.getLogger(__name__)
router = APIRouter()


async def get_current_user_websocket(websocket: WebSocket, token: str = Query(...)):
    """
    Get current user from WebSocket query parameter token
    """
    try:
        user_data = await verify_token_with_auth_server(token)
        if not user_data:
            await websocket.close(code=4001, reason="Invalid token")
            return None
        
        return user_data
    except Exception as e:
        logger.error(f"WebSocket authentication failed: {e}")
        await websocket.close(code=4001, reason="Authentication failed")
        return None


@router.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket, token: str = Query(...)):
    """
    Main WebSocket endpoint for real-time execution updates
    
    Usage: ws://localhost:8003/api/fs/ws?token=<access_token>
    """
    
    # Authenticate user
    user_data = await get_current_user_websocket(websocket, token)
    if not user_data:
        return
    
    user_id = user_data.get("id", "unknown")
    
    # Connect user
    await connection_manager.connect(websocket, user_id)
    
    try:
        while True:
            # Wait for messages from client
            data = await websocket.receive_text()
            
            try:
                message = json.loads(data)
                await handle_websocket_message(websocket, user_id, message)
                
            except json.JSONDecodeError:
                await connection_manager.send_personal_message({
                    "type": "error",
                    "message": "Invalid JSON format"
                }, websocket)
            except Exception as e:
                logger.error(f"Error handling WebSocket message: {e}")
                await connection_manager.send_personal_message({
                    "type": "error",
                    "message": "Failed to process message"
                }, websocket)
                
    except WebSocketDisconnect:
        connection_manager.disconnect(websocket, user_id)
    except Exception as e:
        logger.error(f"WebSocket error for user {user_id}: {e}")
        connection_manager.disconnect(websocket, user_id)


async def handle_websocket_message(websocket: WebSocket, user_id: str, message: dict):
    """
    Handle incoming WebSocket messages from client
    """
    message_type = message.get("type")
    
    if message_type == "subscribe_execution":
        # Subscribe to execution updates
        execution_id = message.get("execution_id")
        if execution_id:
            connection_manager.subscribe_to_execution(user_id, execution_id)
            await connection_manager.send_personal_message({
                "type": "subscription_confirmed",
                "execution_id": execution_id,
                "message": f"Subscribed to execution {execution_id}"
            }, websocket)
        else:
            await connection_manager.send_personal_message({
                "type": "error",
                "message": "execution_id is required for subscription"
            }, websocket)
    
    elif message_type == "unsubscribe_execution":
        # Unsubscribe from execution updates
        execution_id = message.get("execution_id")
        if execution_id:
            connection_manager.unsubscribe_from_execution(user_id, execution_id)
            await connection_manager.send_personal_message({
                "type": "unsubscription_confirmed",
                "execution_id": execution_id,
                "message": f"Unsubscribed from execution {execution_id}"
            }, websocket)
    
    elif message_type == "ping":
        # Heartbeat/ping message
        await connection_manager.send_personal_message({
            "type": "pong",
            "timestamp": message.get("timestamp")
        }, websocket)
    
    elif message_type == "get_stats":
        # Get connection statistics (for debugging)
        stats = connection_manager.get_connection_stats()
        await connection_manager.send_personal_message({
            "type": "stats",
            "data": stats
        }, websocket)
    
    else:
        await connection_manager.send_personal_message({
            "type": "error",
            "message": f"Unknown message type: {message_type}"
        }, websocket)


@router.get("/ws/stats")
async def get_websocket_stats():
    """
    Get WebSocket connection statistics
    """
    return connection_manager.get_connection_stats()